### PR TITLE
Hooks: Allow slashes in the namespace

### PR DESCRIPTION
**Pull request by gziolo**
We would like to use slashes in the hook's namespace as follows:
```js
addFilter( 'registerBlockType', 'core/custom-class-name/attribute', addAttribute );
addFilter( 'BlockEdit', 'core/custom-class-name/inspector-control', addInspectorControl );
addFilter( 'getSaveContent.extraProps', 'core/custom-class-name/save-props', addSaveProps );
```

It is not possible at the moment. This PR changes it and relaxed validation to allow slashes, too.

### CODE REVIEW

**Comment by adamsilverstein** [comment]
> Should we disallow backslash \ for the issues it had presented in behaving as part of an escape sequence?

I think backslash is disallowed with the current check.

**Comment by gziolo** [comment]
> I think backslash is disallowed with the current check.

I will double check before merging. From my own experience from the past, it is not allowed, but @aduth found some magical use case where it doesn't complain.

File: packages/hooks/src/validateNamespace.js [hunk 1]
```diff
@@ -13,8 +13,8 @@ function validateNamespace( namespace ) {
 		return false;
 	}
 
-	if ( ! /^[a-zA-Z][a-zA-Z0-9_.\-]*$/.test( namespace ) ) {
-		console.error( 'The namespace can only contain numbers, letters, dashes, periods and underscores.' );
+	if ( ! /^[a-zA-Z][a-zA-Z0-9_.\-/]*$/.test( namespace ) ) {
```
Comment (aduth) [inline packages/hooks/src/validateNamespace.js]: The forward slash needs to be escaped `\/`.

**Comment by aduth** [comment]
>I think backslash is disallowed with the current check.

Yeah, you're right, I was thrown off by the escaping of dash `-`, but also that we already use a `core\foo` convention in Gutenberg? Why would that have been allowed previously, but not changed here?

**Comment by adamsilverstein** [comment]
@aduth are you sure you are pulling in the latest version in Gutenberg with `@next`? Otherwise you may not be using the latest version from npm. Probably time to publish a release :)

File: packages/hooks/src/test/index.test.js [hunk 2]
```diff
@@ -193,6 +193,13 @@ test( 'cannot add filters with invalid namespaces', () => {
 	);
 } );
 
+test( 'cannot add filters with namespace containing backslash', () => {
+	addFilter( 'hook_name', 'i\n\v\a\l\i\d\a\n\a\m\e', () => null );
```
Comment (gziolo) [inline packages/hooks/src/test/index.test.js]: Fun fact is that `invalidn\amae` will pass. I have no idea how to fix that. We can open a follow-up if we are really concerned about it.

**Pull request merged**